consensus/istanbul/backend/announce.go

```diff
@@ -428,12 +417,58 @@ func (sb *Backend) generateAnnounce(version uint) (*istanbul.Message, common.Add
 	// Sign the announce message
 	if err := msg.Sign(sb.Sign); err != nil {
 		logger.Error("Error in signing an Announce Message", "AnnounceMsg", msg.String(), "err", err)
-		return nil, common.Address{}, err
+		return nil, err
 	}
 
 	logger.Debug("Generated an announce message", "IstanbulMsg", msg.String(), "AnnounceData", announceData.String())
 
-	return msg, msg.Address, nil
+	return msg, nil
+}
+
+// This is a helper function for generateAnnounce.
+// Returns the announce records intended for validators whose entries in the
+// val enode table do not exist or are outdated.
+func (sb *Backend) generateAnnounceRecords(enodeURL string) ([]*announceRecord, error) {
+	allSignedAnnVersionEntries, err := sb.signedAnnounceVersionTable.GetAll()
+	if err != nil {
+		return nil, err
+	}
+	allValEnodes, err := sb.valEnodeTable.GetAllValEnodes()
+	if err != nil {
+		return nil, err
+	}
+	var announceRecords []*announceRecord
+	for _, signedAnnVersionEntry := range allSignedAnnVersionEntries {
+		// Don't generate an announce record for ourselves
+		if signedAnnVersionEntry.Address == sb.Address() {
+			continue
+		}
+
+		valEnode := allValEnodes[signedAnnVersionEntry.Address]
+		// If the version in the val enode table is up to date with the version
+		// we are aware of (or is newer in the case the remote validator sent us
```

Comment:
Nit: Could reword the comment in the parenthesis. I am having trouble understanding what you are trying to say.